### DISCUSSION TITLE
[RFC] New Component: ChatBox

### DISCUSSION THREAD

**Discussion by MadCcc** [ant-design/ant-design#48374] [RFCs]
## Motivation

ChatBox is generally used in AI Chat scenarios, providing ability to show formatted message with streaming motion.

![image](https://github.com/ant-design/ant-design/assets/27722486/8556545b-0ecb-42d2-8b00-5934ce31d345)

## API

| Props | Descriptions | Type | Default |
| --- | --- | --- | --- |
| avatar | Avatar component | `ReactNode` | - |
| placement | Direction of Message | `'start' \| 'end'` | - |
| loading | Loading state | `boolean` |  | - |
| contentRender | Display cuztomized content | `(content: string) => ReactNode` | - |
| typing | Show message with stepping motion | `boolean \| { step: number, interval: number }` | false |
| content | Content of ChatBox | `string` | - |

### Basic

![image](https://github.com/ant-design/ant-design/assets/27722486/50e705f7-dda6-4e2b-bb6c-e4f0b4e7280e)

```tsx
export default () => {
  return (
    <ChatBox content="Lorem ipsum dolor sit amet, consectetur adipiscing elit. ..." />
  );
}
```

### Waiting for Response

![image](https://github.com/ant-design/ant-design/assets/27722486/fde1a08f-b625-49e8-ab50-c95e500816e7)

```tsx
export default () => {
  return (
    <ChatBox loading="生成中，请稍候..." />
  );
}
```

- Loading bar should refer nprogress - it will never reach the end, unless `loading` is set false manually.

### Step

![image](https://github.com/ant-design/ant-design/assets/27722486/63f0a4eb-ad40-450f-9078-103a471673d1)

```tsx
export default () => {
  const [value, setValue] = useState('');

  useEffect(() => {
    setInterval(() => {
      setValue(prev => prev + 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.');
    }, 1000);
  }, [])

  return (
    <ChatBox 
      content={value} 
      step={{ step: 20, interval: 500 }}
    />
  );
}
```

- `step: 20` means 20 charactes will be append in each stream chunk.
- `interval: 500` means new chunk will wait 500ms after previous chunk is rendered.

### With Markdown

```tsx
import markdownit from 'markdown-it'
const md = markdownit();

export default () => {
  return (
    <ChatBox 
      content="# Title  \nLorem ipsum dolor sit amet, consectetur adipiscing elit. ..." 
      contentRender={(content) => <div dangerouslySetInnerHTML={{ __html: md.render(content)}} />}
    />
  );
}
```
- `content` of `render` only contains stream content if `stream` is enabled.

## Detailed Design

![image](https://github.com/ant-design/ant-design/assets/27722486/4e157332-09fb-4aa7-9d83-f41893149284)
![image](https://github.com/ant-design/ant-design/assets/27722486/dfc1bea6-522d-49be-84e1-55a44a82505f)

**Comment by li-jia-nan** [comment]
🎉🎉🎉

**Comment by crazyair** [comment]
要支持 Markdown ？

**Reply by MadCcc** [reply]
用户自己集成

**Comment by kiner-tang** [comment]
建议增加一种报错状态，gpt 聊天啥的，尤其是生成图片之类的，经常会有失败的情况，失败后可以加一个重试按钮等

**Reply by afc163** [reply]
把渲染区域预留出来就好，作为原子组件，先不设计太多交互。

**Comment by crazyair** [comment]
```tsx
<ChatBox  
    items={[
       { content: 你好 , placement: start }
       { content: 你是谁 , placement: 'end'}
    ]}
    footer={<Input/>}
/>
``

**Reply by MadCcc** [reply]
可以顺序堆叠多个 ChatBox 来实现

**Comment by afc163** [comment]
把内容放 children 里？

**Reply by MadCcc** [reply]
children 的类型不只是 string，希望是用户提供 raw text

**Comment by kiner-tang** [comment]
感觉可以增加增加一个消息时间的展示，可以展示在卡片顶部或底部

**Reply by MadCcc** [reply]
后面考虑开放这些 slot 吧

**Comment by afc163** [comment]
这个 loading 在没办法拿到具体的进度时就没法用了，需要提供一个普适的版本。

**Reply by MadCcc** [reply]
做一个 nprogress 那种无限 loading 就好了，前端的 loading 基本都是假的

**Reply by afc163** [reply]
这种 loading 会给用户错误的预期的，得有一个普通的 loading。

**Reply by afc163** [reply]
现在这个设计应该是 `progress={{ precent: 0.1 }}`，loading 就是一个 loading 图标。

类似：<img width="179" alt="截屏2024-04-11 13 48 43" src="https://github.com/ant-design/ant-design/assets/507615/d75e68a3-3b28-4ea7-8f57-f3ec01c2f62d">

**Reply by MadCcc** [reply]
这个和我们设计里的 loading 在使用场景上有啥区别？

**Reply by afc163** [reply]
设计的 loading 不实用，没具体进度数值的场景用不起来的。

**Reply by li-jia-nan** [reply]
这个怎么说？要改成什么样子？

**Reply by afc163** [reply]
这个的实现就是用 `progress={{ precent: 0.1 }}`，loading 另行设计和实现。

**Reply by li-jia-nan** [reply]
这个等周一上班了再和豆酱闲夕讨论一下

**Comment by li-jia-nan** [comment]
- #48384
- https://github.com/ant-design/x/pull/19

**Comment by afc163** [comment]
ChatBox => Chat.Bubble

**Reply by li-jia-nan** [reply]
要拆两个组件吗？

**Reply by MadCcc** [reply]
这是要把 AI 相关组件都收到 Chat 里？
我现在其实不太推荐用 Chat.Bubble 的形式，会影响 RSC 使用

**Reply by afc163** [reply]
后面还会不会有别的组件，我也没想清楚，你和豆酱聊聊？

**Comment by kiner-tang** [comment]
我们这个东西的定位一定是 AI 场景吗？感觉类似 ChatBox 的组件，在 IM 等聊天场景也是可以呀！是不是不要把组件限死了

**Reply by afc163** [reply]
没有限定吧。
